group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,448 @@
+            if (context.header.apiVersion() >= 9) {
+                throw Errors.STALE_MEMBER_EPOCH.exception();
+            } else {
+                throw Errors.UNSUPPORTED_VERSION.exception();
+            }
+        }
+    }
+
+    /**
+     * Handles an OffsetCommit request.
+     *
+     * @param context The request context.
+     * @param request The OffsetCommit request.
+     *
+     * @return A Result containing the OffsetCommitResponseData response and
+     *         a list of records to update the state machine.
+     */
+    public CoordinatorResult<OffsetCommitResponseData, Record> commitOffset(
+        RequestContext context,
+        OffsetCommitRequestData request
+    ) throws ApiException {
+        validateOffsetCommit(context, request);
+
+        final long currentTimeMs = time.milliseconds();
+        // "default" expiration timestamp is defined as now + retention. The retention may be overridden
+        // in versions from v2 to v4. Otherwise, the retention defined on the broker is used. If an explicit
+        // commit timestamp is provided (v1 only), the expiration timestamp is computed based on that.
+        final OptionalLong expireTimestampMs = request.retentionTimeMs() == OffsetCommitRequest.DEFAULT_RETENTION_TIME ?
+            OptionalLong.empty() : OptionalLong.of(currentTimeMs + request.retentionTimeMs());
+        final OffsetCommitResponseData response = new OffsetCommitResponseData();
+        final List<Record> records = new ArrayList<>();
+
+        request.topics().forEach(topic -> {
+            final OffsetCommitResponseData.OffsetCommitResponseTopic topicResponse =
+                new OffsetCommitResponseData.OffsetCommitResponseTopic()
+                    .setName(topic.name());
+            response.topics().add(topicResponse);
+
+            topic.partitions().forEach(partition -> {
+                if (partition.committedMetadata() != null && partition.committedMetadata().length() > offsetMetadataMaxSize) {
+                    topicResponse.partitions().add(new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.OFFSET_METADATA_TOO_LARGE.code()));
+                } else {
+                    log.debug("[GroupId {}] Committing offsets {} for partition {}-{} from member {} with leader epoch {}.",
+                        request.groupId(), partition.committedOffset(), topic.name(), partition.partitionIndex(),
+                        request.memberId(), partition.committedLeaderEpoch());
+
+                    topicResponse.partitions().add(new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.NONE.code()));
+
+                    final OptionalInt leaderEpoch = partition.committedLeaderEpoch() == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
```

Comment:
This one is from [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala#L480).

Reply:
this and `commitTimestampMs` below have these checks based on a static value. this also happens in OffsetAndMetadata#fromRecord. i'm wondering if we can locate them together

Reply:
Reworked this part. Let me know if it works for your.

Reply:
looks much better. thanks!